wallet-api/src/Wallet/Emulator/AddressMap.hs

```diff
@@ -1 +1,111 @@
-module Wallet.Emulator.AddressMap where
\ No newline at end of file
+{-# LANGUAGE RecordWildCards #-}
+module Wallet.Emulator.AddressMap(
+    AddressMap(..),
+    empty,
+    addAddress,
+    addAddresses,
+    values,
+    fromTxOutputs,
+    knownAddresses,
+    updateAddresses,
+    getUtxoAt,
+    setUtxoAt
+    ) where
+
+import           Data.Map       (Map)
+import qualified Data.Map       as Map
+import           Data.Maybe     (catMaybes)
+import           Data.Monoid    (Monoid (..), Sum (..))
+import           Data.Semigroup (Semigroup (..))
+import qualified Data.Set       as Set
+
+import           Wallet.UTXO    (Address', Tx (..), TxIn (..), TxIn', TxOut (..), TxOutRef (..), TxOutRef', Value,
+                                 hashTx)
+
+-- | A map of [[Address']]es and their unspent outputs
+newtype AddressMap = AddressMap { getAddressMap :: Map Address' (Map TxOutRef' Value) }
+    deriving Show
+
+instance Semigroup AddressMap where
+    (AddressMap l) <> (AddressMap r) = AddressMap (Map.unionWith add l r) where
+        add = Map.unionWith (+)
+
+instance Monoid AddressMap where
+    mappend = (<>)
+    mempty = empty
+
+-- | The unspent transaction outputs at an address
+getUtxoAt :: Address' -> AddressMap -> Map TxOutRef' Value
+getUtxoAt addr = Map.findWithDefault Map.empty addr . getAddressMap
+
+-- | Set the funds at an address to a value
+setUtxoAt :: Address' -> Map TxOutRef' Value -> AddressMap -> AddressMap
```

Comment:
Maybe these two should be a lens?

Reply:
I turned them into a lens by making `AddressMap` in instance of `At` and `Ixed`. This required the addition of a dep. on `microlens-ghc` though. If you want me to change it back please let me know.